ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   * @param rackAware if {@code true}, only return disks in nodes that do not share racks with the nodes
+   *                  in {@code dataNodesUsed}
+   * @param numDisks how many disks to return in the sample
+   * @return a list of {@link Disk}s of length {@code numDisks}, or fewer if {@code numDisks} valid samples could not
+   *         be found
+   */
+  private List<Disk> getRandomDiskCandidateSample(List<DataNode> dataNodes, Set<DataNode> dataNodesUsed,
+      long replicaCapacityInBytes, boolean rackAware, int numDisks) {
+    Set<Long> rackIdsUsed = new HashSet<>();
+    if (rackAware) {
+      for (DataNode dataNode : dataNodesUsed) {
+        rackIdsUsed.add(dataNode.getRackId());
+      }
+    }
+    List<Disk> diskCandidates = new ArrayList<>();
+    int selectionBound = dataNodes.size();
+    Random randomGen = new Random();
+    while ((selectionBound > 0) && (diskCandidates.size() < numDisks)) {
+      int selectionIndex = randomGen.nextInt(selectionBound);
+      DataNode candidate = dataNodes.get(selectionIndex);
+      if (!dataNodesUsed.contains(candidate) && !rackIdsUsed.contains(candidate.getRackId())) {
+        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(candidate, replicaCapacityInBytes);
+        if (diskWithMostCapacity != null) {
+          diskCandidates.add(diskWithMostCapacity);
+        }
+      }
+      selectionBound--;
+      Collections.swap(dataNodes, selectionIndex, selectionBound);
+    }
+    return diskCandidates;
+  }
+
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}.  Does not retry if fewer than
+   * {@code replicaCountPerDatacenter} disks cannot be allocated.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param rackAware if {@code true}, attempt a rack-aware allocation
+   * @return A list of {@link Disk}s
+   */
   private List<Disk> allocateDisksForPartition(int replicaCountPerDatacenter, long replicaCapacityInBytes,
-      List<Datacenter> dataCenters) {
-    ArrayList<Disk> allocatedDisks = new ArrayList<Disk>();
+      Datacenter datacenter, boolean rackAware) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    Set<DataNode> nodesToExclude = new HashSet<>();
+    List<DataNode> dataNodes = new ArrayList<>(datacenter.getDataNodes());
+    for (int i = 0; i < replicaCountPerDatacenter; i++) {
+      List<Disk> diskCandidates =
+          getRandomDiskCandidateSample(dataNodes, nodesToExclude, replicaCapacityInBytes, rackAware, NUM_CHOICES);
+      if (diskCandidates.size() > 0) {
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
```

Comment:
A subtle difference from the referenced algorithm is that in our use case, the events are *dependent* - that is here we *exclude* the candidate (which we have to, because no more than one replica can be on a node). In the referenced example, the chosen candidates *remain* in the candidate list - the chosen bin for a ball is eligible to be chosen for subsequent balls. I'd bet that the dependence makes some difference as far as the point of the article is concerned, but perhaps not too much.